server/src/main/java/org/elasticsearch/common/util/SerializableBytesRefHash.java

```diff
@@ -0,0 +1,214 @@
+
+        try {
+            ids = bigArrays.newLongArray(capacity, true);
+            for (int i = 0; i < capacity; ++i) {
+                ids.set(i, in.readVLong());
+            }
+
+            // there is just 1 way of serializing at the moment
+            SerializationType serializationType = SerializationType.fromType(in.readByte());
+            if (SerializationType.RAW_BUFFERS.equals(serializationType) == false) {
+                throw new IllegalArgumentException("unsupported type");
+            }
+
+            // startOffsets
+            long sizeOfStartOffsets = in.readVLong();
+            startOffsets = bigArrays.newLongArray(sizeOfStartOffsets, true);
+            for (int i = 0; i < sizeOfStartOffsets; ++i) {
+                startOffsets.set(i, in.readVLong());
+            }
+
+            // bytes
+            long sizeOfBytes = in.readVLong();
+            bytes = bigArrays.newByteArray(sizeOfBytes, true);
+
+            for (int i = 0; i < sizeOfBytes; ++i) {
+                bytes.set(i, in.readByte());
+            }
+
+            // hashes
+            long sizeOfHashes = in.readVLong();
+            hashes = bigArrays.newIntArray(sizeOfHashes, true);
+            for (int i = 0; i < sizeOfHashes; ++i) {
+                hashes.set(i, in.readZInt());
+            }
+
+            int seed = in.readInt();
+            hash = new SerializableBytesRefHash(capacity, maxLoadFactor, size, maxSize, ids, startOffsets, bytes, hashes, seed, bigArrays);
+            success = true;
+        } finally {
+            if (success == false) {
+                Releasables.close(ids, startOffsets, bytes, hashes);
+            }
+        }
+
+        return hash;
+    }
+
+    @Override
+    public long find(BytesRef key) {
+        return find(key, hash(key));
+    }
+
+    @Override
+    public long add(BytesRef key) {
+        return add(key, hash(key));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeFloat(maxLoadFactor);
+        out.writeVLong(size);
+        out.writeVLong(maxSize);
+
+        long capacity = capacity();
+        out.writeLong(capacity);
+
+        for (int i = 0; i < capacity; ++i) {
+            out.writeVLong(ids.get(i));
+        }
+
+        out.write(SerializationType.RAW_BUFFERS.getType());
+
+        long sizeOfStartOffsets = startOffsets.size();
+        long sizeOfBytes = bytes.size();
+        long sizeOfHashes = hashes.size();
+
+        // startOffsets
+        out.writeVLong(sizeOfStartOffsets);
+        for (int i = 0; i < sizeOfStartOffsets; ++i) {
+            out.writeVLong(startOffsets.get(i));
+        }
+
+        // bytes
+        out.writeVLong(sizeOfBytes);
+        for (int i = 0; i < sizeOfBytes; ++i) {
+            out.writeByte(bytes.get(i));
+        }
+
+        // hashes
+        out.writeVLong(sizeOfHashes);
+
+        for (int i = 0; i < sizeOfHashes; ++i) {
+            out.writeZInt(hashes.get(i));
```

Comment:
I am not sure it's a good idea to use `writeZInt` here.  Looking at how it works, it takes 5 bytes to send a value where bits 28-31 are non-zero.  If the hash function is good then there's only a 1 in 16 chance of the 4 most significant bits all being zero.  So 15 out of 16 times this will send 5 bytes.  It seems like it would be more efficient to just use `writeInt` instead.